Fetch the language revision file on bootstrapping

southp opened the pull request:
## Summary
This PR is based on https://github.com/Automattic/wp-calypso/pull/26973. It introduces a new server-side module, `i18n-bootstrap`, which contains the function `getLangRevisions()` for fetching the language revision map file along side `user-bootstrap`.

## Test Plan
~1. Refer to code-D17741 for instructions of the backend preparation.~ This is no longer needed since it has been deployed.

1. `npm run build-docker`
1. `npm run docker`
1. Access your dockerized calypso according to your host setup. In my case, it is http://wpcalypso.wordpress.com.
1. Open the HTML inspector. The language file should have the revision number appended as the cache buster. e.g. `<script src="//widgets.wp.com/languages/calypso/zh-tw.js?v=35623"></script>`
1. In your browser console, execute `window.langRevisions` should print the revision map, indicating it has been inlined properly.
1. Switch your interface language and make sure the above stay true.
1. Undo the backend preparation from step 1, and see that the language file is loaded without the cache buster.

southp commented on server/pages/index.js:
```diff
@@ -390,14 +401,32 @@ function setUpLoggedInRoute( req, res, next ) {
 			} );
+
+		Promise.all( [ langPromise, userPromise ] ).then( results => {
+			const rejectedWithErrors = results.filter( r => null != r && r.error );
```
@ockham I especially need your help here. This is just a demonstration of my rough idea and I'm not even sure if it worked or not.

When there is only `user-bootstrap`, it is `next( error )` if something goes wrong. What would we do with the `error` passed in and where is the code handling it? Now that there are multiple async computations that could go wrong, how do you envision us handling them?

ockham replied:
Good questions! 

> What would we do with the `error` passed in and where is the code handling it?

Passing an object to `next` will cause our server-side router -- Express -- to fall through to an error handling middleware, which is defined by taking 4 arguments `( err, req, res, next )` rather than the usual 3 `( req, res, next )`. In our case, that middleware is defined here:

https://github.com/Automattic/wp-calypso/blob/fcaba1e63aa6cb802cc1842e97fdde1ac4f50e56/server/pages/index.js#L498-L513

ockham replied:
> This is just a demonstration of my rough idea and I'm not even sure if it worked or not.
> [...]
> Now that there are multiple async computations that could go wrong, how do you envision us handling them?

I haven't honestly used `Promise.all( [ langPromise, userPromise ] )` myself :sweat_smile: 

Looking at the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all), it seems to me that we'd only enter the `.then()` clause if all promises in the list are fulfilled -- which would mean that `rejectedWithErrors` would always be empty.

If I'm reading those docs correctly, we need a `catch()` clause instead:

> It rejects with the reason of the first promise that rejects.

ockham replied:
I think we would call a normal `next()` in the `.then()` clause, and a `next( err )` in the `.catch()` clause.

southp replied:
To think about it more, I think you are right that I have gone too far. I've simplified it in af8d712 so all the Promise objects simply `throw` the errors and let the `Promise.all()` catch function do the rest.

ockham commented on client/lib/i18n-utils/switch-locale.js:
```diff
@@ -11,14 +11,41 @@ import { noop } from 'lodash';
+ *
+ * @return {String} The path URL to the language files.
+ */
+export function languageFilePathUrl() {
```
Maybe prepend `get` to the function name?

ockham replied:
(To indicate it returns something, and has no side effects.)

ockham commented on client/lib/i18n-utils/switch-locale.js:
```diff
@@ -11,14 +11,41 @@ import { noop } from 'lodash';
+ *
+ * @return {String} A language file URL.
+ */
+export function languageFileUrl( localeSlug, fileType = 'json' ) {
```
Maybe prepend `get` to the function name? Maybe prepend get to the function name?

ockham commented on client/lib/i18n-utils/switch-locale.js:
```diff
@@ -11,14 +11,41 @@ import { noop } from 'lodash';
+	}
+
+	const revision = getLangRevision( localeSlug );
+	const fileUrl = languageFilePathUrl() + `${ localeSlug }.${ fileType }`;
```
The mix of string concatenation and template literals is a bit awkward -- let's maybe do

```js
const fileUrl = `${ languageFilePathUrl() }${ localeSlug }.${ fileType }`;
```

ockham commented on client/lib/i18n-utils/test/switch-locale.js:
```diff
@@ -0,0 +1,64 @@
+			hasMockedWindow = true;
+		}
+
+		expect( languageFileUrl( 'ja' ).substring( 0, 2 ) ).toEqual( '//' );
```
You could use lodash's `startsWith()` here.

ockham commented on server/i18n-bootstrap/index.js:
```diff
@@ -0,0 +1,22 @@
+
+/**
+ * Internal dependencies
+ */
```
The comment block doesn't really apply -- this isn't really a dependency :slightly_smiling_face:

ockham commented on server/i18n-bootstrap/index.js:
```diff
@@ -0,0 +1,22 @@
+const LANG_REVISION_FILE_URL = 'https://widgets.wp.com/languages/calypso/lang-revisions.json';
+
+export const fetchLangRevisions = () => {
+	return new Promise( ( resolve, reject ) => {
```
Do we even need the `Promise` wrapper? Seems like the result from `superagent.get()` [should be `then`able](https://visionmedia.github.io/superagent/#request-basics), so I think I'd suggest dropping this file altogether, and directly using

```js
superagent.get( LANG_REVISION_FILE_URL )
	.then( /*...*/ )
	.catch( /*...*/ );
```

in `server/pages/index.js`.

ockham commented on client/lib/i18n-utils/utils.js:
```diff
@@ -174,3 +174,33 @@ export function getForumUrl( localeSlug = getLocaleSlug() ) {
 	return `//en.forums.wordpress.com`;
 }
+
+let langRevisions = {};
```
Hmm. I'm not a big fan of storing things like this at module level. Among other things, this isn't great for testing (see the required mocks). Furthermore, it doesn't really work for server-side rendering, since it would pollute across requests there (which is why I like to isolate things at request context level) -- BTW this is one of the main reasons why we can't use `i18n-calypso` for server-side rendering yet.

Can we maybe store this somehow at context level, or in Redux state instead?

southp replied:
That's a great point. The client side would need some extra refactoring work and could bloat this PR, so I've separated it out to https://github.com/Automattic/wp-calypso/pull/27179/ and keep the server-side bits here.

ockham commented on server/pages/index.js:
```diff
@@ -388,14 +400,33 @@ function setUpLoggedInRoute( req, res, next ) {
+		} );
 	} else {
-		next();
+		fetchLangRevisions()
```
Can we avoid the duplication of `fetchLangRevisions()` here? Maybe by creating a promises array to contain `langPromise` before the `if ( config.isEnabled( 'wpcom-user-bootstrap' ) ) {` conditional, and inside that conditional, appending `userPromise` to the array. Then finally, we'd need to move the `Promises.all()` thing below that conditional.

ockham commented:
Thanks for working on this, this is looking quite good already!

I left a few comments -- a lot of which are merely minor code nitpicks, feel free to ignore those.

There two areas that would be great to address:
* Storing lang revisions at module level
* The promises and conditionals in `server/pages`. I left a few notes there, I hope they make sense -- otherwise, feel free to ping me with questions!

southp commented on client/lib/i18n-utils/switch-locale.js:
```diff
@@ -11,14 +11,41 @@ import { noop } from 'lodash';
+export function languageFileUrl( localeSlug, fileType = 'json' ) {
+	if ( -1 === [ 'js', 'json' ].indexOf( fileType ) ) {
+		fileType = 'json';
+	}
```
Yes, it is on my list for future iteration :)

southp commented:
@ockham Thanks for the thorough review! I've addressed most of your feedback. Some major ones:

1. Remove the stateful language revision series function and store the language revision map elsewhere. For the server side, it's in the request context. For the client side, I plan to store it in redux and it will be implemented in the other PR: https://github.com/Automattic/wp-calypso/pull/27179/
1. Remove `i18n-bootstrap` all together.
1. Eliminate code duplication due to `user-bootstrap` feature flag. Great suggestion!

I didn't update the `then()` handler as explained in [this a-bit-too-long explanation](https://github.com/Automattic/wp-calypso/pull/26975#discussion_r217317119). If it doesn't make sense, I'm happy to iterate further.

When you have some time, I'd be grateful if you can have a revisit :)

ockham commented on server/pages/index.js:
```diff
@@ -388,15 +405,23 @@ function setUpLoggedInRoute( req, res, next ) {
 	}
+
+	Promise.all( setupRequests ).then( results => {
+		const rejectedWithErrors = results.filter( r => null != r && r.error );
```
We avoid type-agnostic comparisons in Calypso, so this should be `r !== null` (we don;t care so much about Yoda case, thanks to type-aware comparisons).

Or you lodash's `get`: `r => get( r, [ 'error' ] )`

Better yet, I think you can even drop the `get` and combine this line with the logic two lines below by using lodash's [`some`](https://lodash.com/docs/4.17.10#some):

```js
if ( some( setupRequests, 'error' ) ) {
	return next( rejectedWithErrors );
}
```

ockham replied:
(I might be mistaken here, resolving the promise early might work just fine, if I'm reading `Promise.all()` docs corectly.)

southp replied:
As the simplification of af8d712, there is no tricky `rejectedWithErrors` anymore.

ockham commented on server/pages/index.js:
```diff
@@ -300,6 +301,24 @@ function setUpLoggedInRoute( req, res, next ) {
+			console.error( 'Failed to fetch the language revision files.', error );
+
+			return error;
+		} );
```
I know I suggested moving `langPromise` up here, but isn't invoking its `.then()` and `.catch()` clauses going to run it immediately? I.e. I think if we don't want to execute it immediately, we can only pass `superagent.get( LANG_REVISION_FILE_URL )` to `Promises.all()`, where the network `response` will be passed as the first arg to the `.then`?

ockham replied:
(I might be mistaken here, resolving the promise early might work just fine, if I'm reading `Promise.all()` docs corectly.)

southp replied:
It's intentional to run it earlier. As you already noted, `Promise.all()` will act as the safe guard here whether the both promises been resolved earlier or not.

ockham commented on server/pages/index.js:
```diff
@@ -388,15 +405,23 @@ function setUpLoggedInRoute( req, res, next ) {
 			} );
-	} else {
-		next();
+
```
([Same as for `langPromise`](https://github.com/Automattic/wp-calypso/pull/26975/files#r217535217) here for `userPromise`.)

ockham commented on server/pages/index.js:
```diff
@@ -322,6 +323,24 @@ function setUpLoggedInRoute( req, res, next ) {
+		.then( response => {
+			const languageRevisions = response.body;
+
+			req.context.languageRevisions = languageRevisions;
```
Do we need to sanitize the network response?

southp replied:
Yes, we do. Thanks for catching this flaw. 🙇 
To not further bloat the size of this PR, I'm going to address this in an up-coming one.

southp replied:
Here it goes: https://github.com/Automattic/wp-calypso/pull/27849
